### ISSUE TITLE
Question Set 2

### ISSUE THREAD

**Issue by andrewniehaus** [andrewniehaus/adleoAN#2]
Q1. This is called translation invariance, and it is generally approached by using convolutions in the model

Q2. A. With spatial dimensions of 28x3, padding of 3 (making the spatial dimension effectively 31x6), 28 weights, 7 channels, 25 pixel kernel size, and stride of 2, the calculation for learnable parameters is ((31*6)(28)(7))/(25*2), which equals about 729 learnable parameters. 

B. with 28 weights and (31*6) pixels, the weights + biases equals 5208 weights and 186 biases. 

C. The computational cost as I understand it would be the number of kernel iterations * the weights / the stride, which would equal about 104. 

Q3. A. With the pooling of a factor of 2 on each dimension, the new dimensions of the feature map would be 14*14*32, which equals 6272. 

B. With the way a convolution calculates every kernel, each pass of an area can be considered its own individual MLP. 

Q4. The trainable parameters in option A are (64*64*100)/25 = 16384
for option B it is 192*10/25 +192 = about 268

for total operations, option A is 64*64*100 = 409600

Option B is 64*64*10/(5*5) = about 1638

the trainable parameters saving for option B is 61X and for total operations is 250X. 

Q5. VGG allows the model to go deeper with comparative computational costs, GoogLeNet has parallel convolutional layers, and ResNet has residual blocks.

**Comment by Rahebe22** [comment]
Thanks for the answers! Please check below for a more complete answers:

Q1-
In image classification, when detecting the presence of a pattern is more important than its specific location, we aim for translation invariance. This means the model should consistently recognize the pattern regardless of its position within the image. Models can achieve this by extracting local features using convolutional layers. These layers apply the same filter across the entire image, focusing on identifying the same feature irrespective of its location (since they use shared weights). This allows the network to recognize the presence of the pattern, not its precise coordinates.
Additionally, pooling layers reduce the spatial resolution while capturing information from local regions, further contributing to translation invariance. Finally, building deeper networks with larger receptive fields allows the model to incorporate a broader context, making it less sensitive to slight location variations and further enhancing translation invariance. Overall, by leveraging these techniques and careful architecture design, models can learn to identify the desired patterns regardless of their specific location in the image, achieving translation invariance. Augmentation is another way to promote translation invariance. Note that this explanation only refers to dense convolution and image classification tasks.

Q2-
b) the shape of the tensor will be: weights: (C_out X f X f X C_in) & biases (C-outX1)

C) the correct answer is: FLOPs = 236,250
Here is how you can calculate it:

[[(f × f × C_in) + [((f × f) − 1) × C_in + (C_in − 1)]] × C_out + C_out] × [(H_in +2p− f) /2+ 1] × [(W_in + 2p – f)/2+ 1]


Q4- check this code:

`
import math

W_in = 64 # Input image width, same as W
H_in = 64 # Input image height, same as H
C_in = 100 # Channels in, # of bands, same as C, same as M
C_out = (192, 10) # Channels out, # of classes, same as D, same as N
f = (5, 1) # dimension of kernel

#What is the number of trainable parameters?
TP1 = ((f[0] * f[0] * C_in) + 1 ) * C_out[0]
TP2 = (((f[1] * f[1] * C_in) + 1 ) * C_out[1]) + (((f[0] * f[0] * C_out[1]) + 1 ) * C_out[0])
no_trainable_parameters_ratio = TP1/TP2
print(f'The number of trainable parameters reduction is {no_trainable_parameters_ratio} %')
FLOPs1 = (((f[0] * f[0] * C_in) + (((f[0] * f[0]) - 1) * C_in + (C_in - 1))) *
C_out[0] + C_out[0]) * (64*64) # since it is same padding th eoutput dimension is 64 by 64

FLOPs2_L1 = (((f[1] * f[1] * C_in) + (((f[1] * f[1]) - 1) * C_in + (C_in - 1))) *
C_out[1] + C_out[1]) * (64*64) #since 1 by 1 conv doesnt change the spatial dimension

FLOPs2_L2 = (((f[0] * f[0] * C_out[1]) + (((f[0] * f[0]) - 1) * C_out[1] + (C_out[1] - 1))) *
C_out[0] + C_out[0]) * (64*64)

FLOPs2 = FLOPs2_L1 + FLOPs2_L2

computation_cost_ratio = FLOPs1/FLOPs2

print (f'The computation cost reduction is {computation_cost_ratio} %')
print (f'TP1: {TP1}')
print (f'TP2: {TP2}')
print (f'FLOPs1: {FLOPs1}')
print (f'FLOPs2: {FLOPs2}')
`
Q5-
VGG used blocks of cov layers with small convolutional filters (3x3) and deep architecture. In VGG each layer is composed of multiple convolutional layers with same padding to maintain the height and weight, followed by a max-pooling layer, which is repeated several times.It also uses smaller strides in the conv layers, which preserves the spatial resolution and results in more accurate object localization in image, as the spatial resolution of the feature maps is preserved. It solved the problem of having limited number of conv layers (log(2)d) since the network couldn't go deeper when dimensions will be consumed in the few conv layers. In a VGG block, conv blocks are grouped in nonlinear transformations which makes no changes in dimensionality. VGG also introduced the idea that deep and narrow networks are better than shallow ones.

GoogleNet introduced a distinction between the stem (data digest), body (data process), and head (prediction) of the CNNs. By concatenating a multibranch convolution block in its body, GoogleNet solved the problem of choosing the right kernel size. Exploring the image over a variety of filters so details at different extents can be recognized.

ResNets introduced the concept of residual connections. The innovation of residual networks is that instead of learning mapping (f(x)=x) directly, they learn the residual mapping (g(x)=f(x)-x). Residual mapping is the difference between the input and output of a layer, so that they are able to solve the problem with the vanishing gradients that can occur in very deep neural networks. In fact, residual connections create shortcut connections to skip layers in the network, allowing the network to learn both the mapping and the residual mapping simultaneously. In this way, the gradient can backpropagat directly to the earlier layers through the shortcut connections, which makes the network to learn deeper representations.
